### ISSUE TITLE
Tile bug?

### ISSUE THREAD

**Issue by Larnoo** [osmdroid/osmdroid#557]
[
![screenshot_2017-02-09-16-04-07](https://cloud.githubusercontent.com/assets/15139682/22774405/ef6026f8-eee1-11e6-86bd-28ff0e4979bd.png)
](url)

The MapView zoom is 4, but the showing tiles contains zoom 5. 

The gray block is a Bitmap, and I can't find where the gray bitmap is from. Because  when I create BitmapDrawable, I draw text at the bitmap.  

My TileSource extends XYTileSource. The code is shown below. Can you help me?

```
public class CKXYTileSource extends XYTileSource {

    public CKXYTileSource(String aName, int aZoomMinLevel, int aZoomMaxLevel, int aTileSizePixels, String aImageFilenameEnding, String[] aBaseUrl) {
        super(aName, aZoomMinLevel, aZoomMaxLevel, aTileSizePixels, aImageFilenameEnding, aBaseUrl);
        BitmapPool.getInstance().clearBitmapPool();
    }


    @Override
    public Drawable getDrawable(final String aFilePath) throws LowMemoryException {
        //Log.d(IMapView.LOGTAG, aFilePath + " attempting to load bitmap");
        try {
            // default implementation will load the file as a bitmap and create
            // a BitmapDrawable from it
            BitmapFactory.Options bitmapOptions = getOptions();
            BitmapPool.getInstance().applyReusableOptions(bitmapOptions);
            Bitmap bitmap = BitmapFactory.decodeFile(aFilePath, bitmapOptions);
            if (bitmap != null) {
                if (!BuildConfig.PRODUCTION) {
                    drawTextOnBitmap(bitmap, aFilePath);
                }
                return new ReusableBitmapDrawable(bitmap);
            } else {
                File bmp = new File(aFilePath);
                if (bmp.exists()) {
                    // if we couldn't load it then it's invalid - delete it
                    Log.d(IMapView.LOGTAG, aFilePath + " is an invalid image file, deleting...");
                    try {
                        new File(aFilePath).delete();
                    } catch (final Throwable e) {
                        Log.e(IMapView.LOGTAG, "Error deleting invalid file: " + aFilePath, e);
                    }
                } else
                    Log.d(IMapView.LOGTAG, "Request tile: " + aFilePath + " does not exist");
            }
        } catch (final OutOfMemoryError e) {
            Log.e(IMapView.LOGTAG, "OutOfMemoryError loading bitmap: " + aFilePath);
            System.gc();
            throw new LowMemoryException(e);
        } catch (final Exception e) {
            Log.e(IMapView.LOGTAG, "Unexpected error loading bitmap: " + aFilePath, e);
            Counters.tileDownloadErrors++;
            System.gc();
        }
        return null;
    }

    @NonNull
    private static BitmapFactory.Options getOptions() {
        BitmapFactory.Options options = new BitmapFactory.Options();
        return options;
    }

    @Override
    public Drawable getDrawable(final InputStream aFileInputStream) throws LowMemoryException {
        try {
            // default implementation will load the file as a bitmap and create
            // a BitmapDrawable from it
            BitmapFactory.Options bitmapOptions = getOptions();
            BitmapPool.getInstance().applyReusableOptions(bitmapOptions);
            final Bitmap bitmap = BitmapFactory.decodeStream(aFileInputStream, null, bitmapOptions);
            if (bitmap != null) {
                if (!BuildConfig.PRODUCTION) {
                    drawTextOnBitmap(bitmap, "tileSource/aFileInputStream/@2x.png.tile");
                }
                return new ReusableBitmapDrawable(bitmap);
            }
        } catch (final OutOfMemoryError e) {
            Log.e(IMapView.LOGTAG, "OutOfMemoryError loading bitmap");
            System.gc();
            throw new LowMemoryException(e);
        }
        return null;
    }

    private void drawTextOnBitmap(Bitmap bitmap, String filePath) {
        Canvas canvas = new Canvas(bitmap);
        Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setColor(Color.RED);
        // text size in pixels
        paint.setTextSize(CKUtil.dip2px(14));
        // draw text to the Canvas center
        Rect bounds = new Rect();
        String tileSource = "tileSource";
        String suffix = "@2x.png.tile";
        int index = filePath.lastIndexOf(tileSource);
        int end = filePath.lastIndexOf(suffix);
        if (index >= 0 && end > index) {
            filePath = filePath.substring(index + tileSource.length(), end);
        }
        paint.getTextBounds(filePath, 0, filePath.length(), bounds);
        canvas.drawText(filePath, 10, bounds.bottom - bounds.top + 10, paint);
        canvas.drawLine(0, 0, 0, bitmap.getHeight(), paint);
        canvas.drawLine(0, bitmap.getHeight(), bitmap.getWidth(), bitmap.getHeight(), paint);
        canvas.drawLine(bitmap.getWidth(), bitmap.getHeight(), bitmap.getWidth(), 0, paint);
        canvas.drawLine(bitmap.getWidth(), 0, 0, 0, paint);
    }
}
```

**Comment by Larnoo** [comment]
@sandykurniawan19

**Comment by spyhunter99** [comment]
@Larnoo are you adding a secondary tile overlay or is this your primary? The secondary tile overalys sometimes have this problem until the map is moved/invalidated.

**Comment by Larnoo** [comment]
I create a default MapView. I don't add secondary tile overlay. @sandykurniawan19
Thank you for your help.

```

  MapView mapView = new Map(context);
  overlayManager = new MyOverlayManager(mapView.getOverlayManager().getTilesOverlay());
  mapView.setOverlayManager(overlayManager);

 ITileSource tileSource = new CKXYTileSource("tileSource", 0, 20, 512, "@2x.png", new String[]{AppConfig.tileServer()});
        mapView.setTileSource(tileSource);
```

**Comment by sandman7920** [comment]
try mapView.setTilesScaledToDpi(true);

**Comment by Larnoo** [comment]
when use this
`mapView.setTilesScaledToDpi(true);`
The probability of this happening  reduced, but it hanppens sometimes.

**Comment by Larnoo** [comment]
Where the gray bitmap is from? Can you explain that？

**Comment by spyhunter99** [comment]
the gray bitmap is the default color in the TilesOverlay

**Comment by pasniak** [comment]
I have a similar problem... I have multiple tile sources, and a listview which controls which tile is displayed. 
```
                                onItemClick {av, v, index, _ ->
                                    val clickedSource = TileSourceFactory.All[index]
                                    map?.setTileSource(clickedSource)
                                }
```
Tile source is switched fine but the map does not really show immediately the new tiles. Only after I start zooming around!  

I tried map.invalidate() and setTilesScaledToDpi but the tiles do not refresh/invalidate.

**Comment by monsieurtanuki** [comment]
@pasniak The memory cache should be cleared. A map `invalidate` isn't enough, it will trigger a display refresh that will first have a look at the cache, where the tile already exist with the previous source/bitmap.

**Comment by pasniak** [comment]
Hmm I stepped into the code. clearTileCache seems to be already happening in 
```
    public void setTileSource(ITileSource aTileSource) {
        this.mTileProvider.setTileSource(aTileSource);
        this.updateTileSizeForDensity(aTileSource);
        this.checkZoomButtons();
        this.setZoomLevel(this.mZoomLevel);
        this.postInvalidate();
    }
```
calling
```
    public void setTileSource(ITileSource aTileSource) {
        super.setTileSource(aTileSource);
        List var2 = this.mTileProviderList;
        synchronized(this.mTileProviderList) {
            Iterator var3 = this.mTileProviderList.iterator();

            while(var3.hasNext()) {
                MapTileModuleProviderBase tileProvider = (MapTileModuleProviderBase)var3.next();
                tileProvider.setTileSource(aTileSource);
                this.clearTileCache();  //<----- `this` is new MapTileProvider; I break here
            }

        }
    }
```

**Comment by pasniak** [comment]
Note that I am in `disconnected from internet but connected to network` mode, would that matter?

**Comment by monsieurtanuki** [comment]
@pasniak I have no access to my laptop now. I'd say though that the cache must be cleared at the `MapTileProviderArray` level, not for each provider.

**Comment by pasniak** [comment]
hmm, i do not use MapTileProviderArray (which would be used for layering maps I understand). 

I have my list of tile sources inheriting from XYTileSource, and I simply call map.setTileSource(clickedSource) to display one map at a time.

**Comment by monsieurtanuki** [comment]
@pasniak If I try the "More Samples / Animations / Animated Marker" and change the "Map Source" (menu button) to "USGS National Map Topo", the source change is immediate. And also when back to "Mapnik".
Is that OK on your device too?

**Comment by monsieurtanuki** [comment]
@Larnoo @pasniak Still a problem?

**Comment by pasniak** [comment]
I have not seen the problem for quite long time.

**Comment by monsieurtanuki** [comment]
Let's assume it's fixed. Feel free to open a new issue if needed.
